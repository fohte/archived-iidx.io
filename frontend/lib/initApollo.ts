import 'isomorphic-fetch'

import { InMemoryCache, NormalizedCacheObject } from 'apollo-cache-inmemory'
import { ApolloClient } from 'apollo-client'
import { ApolloLink } from 'apollo-link'
import { setContext } from 'apollo-link-context'
import { onError } from 'apollo-link-error'
import { HttpLink } from 'apollo-link-http'

import { auth } from '../lib/firebaseApp'
import isBrowser from './isBrowser'

let apolloClient: ApolloClient<NormalizedCacheObject> | null = null

const handleError = onError(({ graphQLErrors, networkError }) => {
  if (graphQLErrors) {
    graphQLErrors.map(({ message, locations, path }) =>
      console.error(
        `[GraphQL error]: Message: ${message}, Location: ${JSON.stringify(
          locations,
        )}, Path: ${path}`,
      ),
    )
  }

  if (networkError) {
    console.error(`[Network error]: ${networkError}`)
  }
})

const withAuthorization = setContext(async (_, { headers }) => {
  if (auth.currentUser) {
    const idToken = await auth.currentUser.getIdToken()

    if (idToken) {
      return {
        headers: { ...headers, authorization: `Bearer ${idToken}` },
      }
    }
  }

  return { headers }
})

const create = (initialState?: NormalizedCacheObject) => {
  const cache = new InMemoryCache().restore(initialState || {})

  const httpLink = new HttpLink({
    uri: process.env.API_URL,
    credentials: 'same-origin',
  })

  return new ApolloClient({
    connectToDevTools: isBrowser,
    ssrMode: !isBrowser,
    link: ApolloLink.from([handleError, withAuthorization, httpLink]),
    cache,
  })
}

const initApollo = (initialState?: NormalizedCacheObject) => {
  // Make sure to create a new client for every server-side request so that data
  // isn't shared between connections (which would be bad)
  if (!isBrowser) {
    return create(initialState)
  }

  // Reuse client on the client-side
  if (!apolloClient) {
    apolloClient = create(initialState)
  }

  return apolloClient
}

export default initApollo
