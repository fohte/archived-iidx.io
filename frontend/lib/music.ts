export const makeTitle = (music: { title: string; subTitle: string }) =>
  `${music.title} ${music.subTitle}`
