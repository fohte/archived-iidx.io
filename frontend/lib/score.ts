import * as _ from 'lodash'

export enum Grade {
  Max = 'MAX',
  A = 'A',
  AA = 'AA',
  AAA = 'AAA',
  B = 'B',
  C = 'C',
  D = 'D',
  E = 'E',
  F = 'F',
}

export const calcScoreRate = (
  score: number,
  numNotes: number,
  precision: number = 2,
) => _.round((score / (numNotes * 2)) * 100, precision)

const getGradeBorders = (numNotes: number): { [key in Grade]: number } => {
  const maxScore = numNotes * 2
  const calcGradeBorder = (coefficients: number): number =>
    Math.ceil(maxScore * coefficients)

  return {
    [Grade.F]: 0,
    [Grade.E]: calcGradeBorder(2 / 9),
    [Grade.D]: calcGradeBorder(3 / 9),
    [Grade.C]: calcGradeBorder(4 / 9),
    [Grade.B]: calcGradeBorder(5 / 9),
    [Grade.A]: calcGradeBorder(6 / 9),
    [Grade.AA]: calcGradeBorder(7 / 9),
    [Grade.AAA]: calcGradeBorder(8 / 9),
    [Grade.Max]: maxScore,
  }
}

export const searchGrade = (
  score: number,
  numNotes: number,
): [Grade | null, number] => {
  if (numNotes * 2 === score) {
    return [Grade.Max, 0]
  }

  const gradeBorders = getGradeBorders(numNotes)

  const diffs = _.mapValues(gradeBorders, border => score - border)
  const diffArray = _.toPairs(diffs) as Array<[Grade, number]>

  const nearest = _.minBy<[Grade, number]>(
    _.filter(diffArray, ([, diff]) => diff >= 0),
    ([, diff]) => Math.abs(diff),
  )

  return nearest || [null, 0]
}

export const searchNextGrade = (
  score: number,
  numNotes: number,
): [Grade | null, number] => {
  if (numNotes * 2 === score) {
    return [Grade.Max, 0]
  }

  const gradeBorders = getGradeBorders(numNotes)

  const diffs = _.mapValues(gradeBorders, border => score - border)
  const diffArray = _.toPairs(diffs) as Array<[Grade, number]>

  const nearest = _.minBy<[Grade, number]>(
    _.filter(diffArray, ([grade, diff]) => diff < 0 && grade !== Grade.F),
    ([, diff]) => Math.abs(diff),
  )

  return nearest || [null, 0]
}
