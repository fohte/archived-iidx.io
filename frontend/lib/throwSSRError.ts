import { ServerResponse } from 'http'

export default (res: ServerResponse | undefined, statusCode: number) => {
  if (res) {
    res.statusCode = statusCode
  }
}
