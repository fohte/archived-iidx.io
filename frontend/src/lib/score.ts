import * as _ from 'lodash'

export enum Grade {
  Max = 'MAX',
  A = 'A',
  AA = 'AA',
  AAA = 'AAA',
  B = 'B',
  C = 'C',
  D = 'D',
  E = 'E',
  F = 'F',
}

export interface GradeDiff {
  grade: Grade | null
  diff: number
}

export const calcScoreRate = (
  score: number,
  numNotes: number,
  precision: number = 2,
) => _.round((score / (numNotes * 2)) * 100, precision)

export const calcGradeBorder = (coefficients: number, numNotes: number) =>
  Math.ceil(numNotes * 2 * coefficients)

export const getGradeBorders = (
  numNotes: number,
): { [key in Grade]: number } => ({
  [Grade.F]: 0,
  [Grade.E]: calcGradeBorder(numNotes, 2 / 9),
  [Grade.D]: calcGradeBorder(numNotes, 3 / 9),
  [Grade.C]: calcGradeBorder(numNotes, 4 / 9),
  [Grade.B]: calcGradeBorder(numNotes, 5 / 9),
  [Grade.A]: calcGradeBorder(numNotes, 6 / 9),
  [Grade.AA]: calcGradeBorder(numNotes, 7 / 9),
  [Grade.AAA]: calcGradeBorder(numNotes, 8 / 9),
  [Grade.Max]: numNotes * 2,
})

export const defaultGradeDiff: GradeDiff = {
  grade: null,
  diff: 0,
}

export const searchGrade = (score: number, numNotes: number): GradeDiff => {
  if (numNotes * 2 === score) {
    return { grade: Grade.Max, diff: 0 }
  }

  const gradeBorders = getGradeBorders(numNotes)

  const diffs = _.mapValues(gradeBorders, border => score - border)
  const diffArray = _.toPairs(diffs) as Array<[Grade, number]>

  const nearest = _.minBy<[Grade, number]>(
    _.filter(diffArray, ([, diff]) => diff >= 0),
    ([, diff]) => Math.abs(diff),
  )

  if (!nearest) {
    return defaultGradeDiff
  }

  return { grade: nearest[0], diff: nearest[1] }
}

export const searchNextGrade = (score: number, numNotes: number): GradeDiff => {
  if (numNotes * 2 === score) {
    return { grade: Grade.Max, diff: 0 }
  }

  const gradeBorders = getGradeBorders(numNotes)

  const diffs = _.mapValues(gradeBorders, border => score - border)
  const diffArray = _.toPairs(diffs) as Array<[Grade, number]>

  const nearest = _.minBy<[Grade, number]>(
    _.filter(diffArray, ([grade, diff]) => diff < 0 && grade !== Grade.F),
    ([, diff]) => Math.abs(diff),
  )

  if (!nearest) {
    return defaultGradeDiff
  }

  return { grade: nearest[0], diff: nearest[1] }
}

export const isMax = ({ grade, diff }: GradeDiff): boolean =>
  grade === Grade.Max && diff === 0
