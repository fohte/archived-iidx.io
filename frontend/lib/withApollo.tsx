import { NormalizedCacheObject } from 'apollo-cache-inmemory'
import { ApolloClient } from 'apollo-client'
import Head from 'next/head'
import React from 'react'
import { ApolloProvider, getDataFromTree } from 'react-apollo'
import { setDisplayName, wrapDisplayName } from 'recompose'

import initApollo from './initApollo'
import isBrowser from './isBrowser'

interface Props {
  serverState: {
    apollo: {
      data: NormalizedCacheObject
    }
  }
}

export default ComposedComponent => {
  class WithData extends React.Component<Props> {
    public static async getInitialProps(ctx) {
      // Initial serverState with apollo (empty)
      let serverState: Props['serverState'] = {
        apollo: {
          data: {},
        },
      }

      // Evaluate the composed component's getInitialProps()
      let composedInitialProps = {}
      if (ComposedComponent.getInitialProps) {
        composedInitialProps = await ComposedComponent.getInitialProps(ctx)
      }

      // Run all GraphQL queries in the component tree
      // and extract the resulting data
      if (!isBrowser) {
        const apollo = initApollo()

        try {
          // Run all GraphQL queries
          await getDataFromTree(
            <ApolloProvider client={apollo}>
              <ComposedComponent {...composedInitialProps} />
            </ApolloProvider>,
            {
              router: {
                asPath: ctx.asPath,
                pathname: ctx.pathname,
                query: ctx.query,
              },
            },
          )
        } catch (error) {
          // Prevent Apollo Client GraphQL errors from crashing SSR.
          // Handle them in components via the data.error prop:
          // https://www.apollographql.com/docs/react/api/react-apollo.html#graphql-query-data-error
        }
        // getDataFromTree does not call componentWillUnmount
        // head side effect therefore need to be cleared manually
        Head.rewind()

        // Extract query data from the Apollo store
        serverState = {
          apollo: {
            data: apollo.cache.extract(),
          },
        }
      }

      return {
        serverState,
        ...composedInitialProps,
      }
    }

    public apollo: ApolloClient<NormalizedCacheObject>

    constructor(props) {
      super(props)
      this.apollo = initApollo(this.props.serverState.apollo.data)
    }

    public render() {
      return (
        <ApolloProvider client={this.apollo}>
          <ComposedComponent {...this.props} />
        </ApolloProvider>
      )
    }
  }

  return setDisplayName(wrapDisplayName(WithData, 'WithData'))(WithData)
}
