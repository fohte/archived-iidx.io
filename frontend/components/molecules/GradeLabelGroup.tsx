import * as React from 'react'

import GradeLabel from '@app/components/atoms/GradeLabel'
import NextGradeLabel from '@app/components/atoms/NextGradeLabel'
import { Grade, searchGrade, searchNextGrade } from '@app/lib/score'

export interface Props {
  score: number
  numNotes: number
}

const isMax = (grade: Grade | null, diff: number): boolean =>
  grade === Grade.Max && diff === 0

const GradeLabelGroup: React.SFC<Props> = ({ score, numNotes }) => {
  const [grade, diff] = searchGrade(score, numNotes)
  const [nextGrade, nextDiff] = searchNextGrade(score, numNotes)

  return (
    <span>
      {grade && <GradeLabel grade={grade} diff={diff} />}
      {nextGrade && !isMax(nextGrade, nextDiff) && (
        <NextGradeLabel grade={nextGrade} diff={nextDiff} />
      )}
    </span>
  )
}

export default GradeLabelGroup
