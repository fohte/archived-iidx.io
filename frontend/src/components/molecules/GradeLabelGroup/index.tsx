import * as React from 'react'

import GradeLabel from '@app/components/atoms/GradeLabel'
import NextGradeLabel from '@app/components/atoms/NextGradeLabel'
import { GradeDiff, isMax } from '@app/lib/score'

export interface Props {
  current: GradeDiff
  next: GradeDiff
}

const GradeLabelGroup: React.SFC<Props> = ({ current, next }) => {
  return (
    <>
      <div style={{ marginBottom: '0.3em' }}>
        <GradeLabel grade={current.grade} diff={current.diff} />
      </div>
      {next.grade && !isMax(next) && (
        <div>
          <NextGradeLabel grade={next.grade} diff={next.diff} />
        </div>
      )}
    </>
  )
}

export default GradeLabelGroup
