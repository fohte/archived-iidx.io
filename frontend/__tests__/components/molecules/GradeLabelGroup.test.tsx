import * as _ from 'lodash'
import * as React from 'react'
import * as renderer from 'react-test-renderer'

import GradeLabelGroup, {
  Props,
} from '@app/components/molecules/GradeLabelGroup'
import { Grade } from '@app/lib/score'

describe('GradeLabelGroup', () => {
  const params: Array<Props & { description: string }> = [
    {
      description: 'when the current grade is MAX-0',
      current: { grade: Grade.Max, diff: 0 },
      next: { grade: Grade.Max, diff: 0 },
    },
    {
      description: 'when the current grade is AAA+100 and MAX-100',
      current: { grade: Grade.AAA, diff: 100 },
      next: { grade: Grade.Max, diff: -100 },
    },
    {
      description: 'when the current grade is AA+100 and AAA-100',
      current: { grade: Grade.AA, diff: 100 },
      next: { grade: Grade.AAA, diff: -100 },
    },
    {
      description: 'when the current grade is A+100 and AA-100',
      current: { grade: Grade.A, diff: 100 },
      next: { grade: Grade.AA, diff: -100 },
    },
    {
      description: 'when the current grade is B+100 and A-100',
      current: { grade: Grade.B, diff: 100 },
      next: { grade: Grade.A, diff: -100 },
    },
    {
      description: 'when the current grade is C+100 and B-100',
      current: { grade: Grade.C, diff: 100 },
      next: { grade: Grade.B, diff: -100 },
    },
    {
      description: 'when the current grade is D+100 and C-100',
      current: { grade: Grade.D, diff: 100 },
      next: { grade: Grade.C, diff: -100 },
    },
    {
      description: 'when the current grade is E+100 and D-100',
      current: { grade: Grade.E, diff: 100 },
      next: { grade: Grade.D, diff: -100 },
    },
    {
      description: 'when the current grade is F+100 and E-100',
      current: { grade: Grade.F, diff: 100 },
      next: { grade: Grade.E, diff: -100 },
    },
    {
      description: 'when the current grade is null',
      current: { grade: null, diff: 0 },
      next: { grade: null, diff: 0 },
    },
  ]

  params.forEach(({ description, current, next }) => {
    it(`renders correctly ${description}`, () => {
      const tree = renderer
        .create(<GradeLabelGroup current={current} next={next} />)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })
})
