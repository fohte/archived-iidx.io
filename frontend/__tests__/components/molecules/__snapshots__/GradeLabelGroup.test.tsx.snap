// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GradeLabelGroup renders correctly when the current grade is A+100 and AA-100 1`] = `
Array [
  <div
    style={
      Object {
        "marginBottom": "0.3em",
      }
    }
  >
    <div
      className="ui blue horizontal label"
      onClick={[Function]}
    >
      A
      <div
        className="detail"
      >
        +100
      </div>
    </div>
  </div>,
  <div>
    <div
      className="ui horizontal label"
      onClick={[Function]}
    >
      AA
      <div
        className="detail"
      >
        -100
      </div>
    </div>
  </div>,
]
`;

exports[`GradeLabelGroup renders correctly when the current grade is AA+100 and AAA-100 1`] = `
Array [
  <div
    style={
      Object {
        "marginBottom": "0.3em",
      }
    }
  >
    <div
      className="ui olive horizontal label"
      onClick={[Function]}
    >
      AA
      <div
        className="detail"
      >
        +100
      </div>
    </div>
  </div>,
  <div>
    <div
      className="ui horizontal label"
      onClick={[Function]}
    >
      AAA
      <div
        className="detail"
      >
        -100
      </div>
    </div>
  </div>,
]
`;

exports[`GradeLabelGroup renders correctly when the current grade is AAA+100 and MAX-100 1`] = `
Array [
  <div
    style={
      Object {
        "marginBottom": "0.3em",
      }
    }
  >
    <div
      className="ui orange horizontal label"
      onClick={[Function]}
    >
      AAA
      <div
        className="detail"
      >
        +100
      </div>
    </div>
  </div>,
  <div>
    <div
      className="ui horizontal label"
      onClick={[Function]}
    >
      MAX
      <div
        className="detail"
      >
        -100
      </div>
    </div>
  </div>,
]
`;

exports[`GradeLabelGroup renders correctly when the current grade is B+100 and A-100 1`] = `
Array [
  <div
    style={
      Object {
        "marginBottom": "0.3em",
      }
    }
  >
    <div
      className="ui black horizontal label"
      onClick={[Function]}
    >
      B
      <div
        className="detail"
      >
        +100
      </div>
    </div>
  </div>,
  <div>
    <div
      className="ui horizontal label"
      onClick={[Function]}
    >
      A
      <div
        className="detail"
      >
        -100
      </div>
    </div>
  </div>,
]
`;

exports[`GradeLabelGroup renders correctly when the current grade is C+100 and B-100 1`] = `
Array [
  <div
    style={
      Object {
        "marginBottom": "0.3em",
      }
    }
  >
    <div
      className="ui black horizontal label"
      onClick={[Function]}
    >
      C
      <div
        className="detail"
      >
        +100
      </div>
    </div>
  </div>,
  <div>
    <div
      className="ui horizontal label"
      onClick={[Function]}
    >
      B
      <div
        className="detail"
      >
        -100
      </div>
    </div>
  </div>,
]
`;

exports[`GradeLabelGroup renders correctly when the current grade is D+100 and C-100 1`] = `
Array [
  <div
    style={
      Object {
        "marginBottom": "0.3em",
      }
    }
  >
    <div
      className="ui black horizontal label"
      onClick={[Function]}
    >
      D
      <div
        className="detail"
      >
        +100
      </div>
    </div>
  </div>,
  <div>
    <div
      className="ui horizontal label"
      onClick={[Function]}
    >
      C
      <div
        className="detail"
      >
        -100
      </div>
    </div>
  </div>,
]
`;

exports[`GradeLabelGroup renders correctly when the current grade is E+100 and D-100 1`] = `
Array [
  <div
    style={
      Object {
        "marginBottom": "0.3em",
      }
    }
  >
    <div
      className="ui black horizontal label"
      onClick={[Function]}
    >
      E
      <div
        className="detail"
      >
        +100
      </div>
    </div>
  </div>,
  <div>
    <div
      className="ui horizontal label"
      onClick={[Function]}
    >
      D
      <div
        className="detail"
      >
        -100
      </div>
    </div>
  </div>,
]
`;

exports[`GradeLabelGroup renders correctly when the current grade is F+100 and E-100 1`] = `
Array [
  <div
    style={
      Object {
        "marginBottom": "0.3em",
      }
    }
  >
    <div
      className="ui black horizontal label"
      onClick={[Function]}
    >
      F
      <div
        className="detail"
      >
        +100
      </div>
    </div>
  </div>,
  <div>
    <div
      className="ui horizontal label"
      onClick={[Function]}
    >
      E
      <div
        className="detail"
      >
        -100
      </div>
    </div>
  </div>,
]
`;

exports[`GradeLabelGroup renders correctly when the current grade is MAX-0 1`] = `
<div
  style={
    Object {
      "marginBottom": "0.3em",
    }
  }
>
  <div
    className="ui red horizontal label"
    onClick={[Function]}
  >
    MAX
    <div
      className="detail"
    >
      0
    </div>
  </div>
</div>
`;

exports[`GradeLabelGroup renders correctly when the current grade is null 1`] = `
<div
  style={
    Object {
      "marginBottom": "0.3em",
    }
  }
>
  <span>
    -
  </span>
</div>
`;
